File: src/common/redis-client/scripts/lua/delete-message.lua
local EQueuePropertyQueueType = ARGV[1]
local EQueuePropertyMessagesCount = ARGV[2]
local EQueuePropertyQueueTypePriorityQueue = ARGV[3]
local EQueuePropertyQueueTypeLIFOQueue = ARGV[4]
local EQueuePropertyQueueTypeFIFOQueue = ARGV[5]
local EMessagePropertyStatus = ARGV[6]
local EMessagePropertyStatusProcessing = ARGV[7]
local EMessagePropertyStatusAcknowledged = ARGV[8]
local EMessagePropertyStatusPending = ARGV[9]
local EMessagePropertyStatusScheduled = ARGV[10]
local EMessagePropertyStatusDeadLettered = ARGV[11]
local EMessagePropertyStatusUnackDelaying = ARGV[12]
local EMessagePropertyStatusUnackRequeuing = ARGV[13]

---

local keyQueueScheduled = ''
local keyQueueDelayed = ''
local keyQueueRequeued = ''
local keyMessage = ''
local keyQueueProperties = ''
local keyQueuePending = ''
local keyQueueDL = ''
local keyQueueAcknowledged = ''
local keyQueuePriorityPending = ''

---

local messageId = ''

---

local keyIndexOffset = 0
local argvIndexOffset = 13

---

local function updateQueue()
    redis.call("HINCRBY", keyQueueProperties, EQueuePropertyMessagesCount, -1)
end

local function deleteMessage()
    local messageStatus = redis.call("HGET", keyMessage, EMessagePropertyStatus)
    if messageStatus == false then
        return 'MESSAGE_NOT_FOUND'
    end
    if messageStatus == EMessagePropertyStatusProcessing then
        return 'MESSAGE_IN_PROCESS'
    end
    local deleted = 0;
    if messageStatus == EMessagePropertyStatusAcknowledged then
        deleted = redis.call("LREM", keyQueueAcknowledged, 1, messageId)
    elseif messageStatus == EMessagePropertyStatusDeadLettered then
        deleted = redis.call("LREM", keyQueueDL, 1, messageId)
    elseif messageStatus == EMessagePropertyStatusScheduled then
        deleted = redis.call("ZREM", keyQueueScheduled, messageId)
    elseif messageStatus == EMessagePropertyStatusUnackDelaying then
        deleted = redis.call("LREM", keyQueueDelayed, 1, messageId)
    elseif messageStatus == EMessagePropertyStatusUnackRequeuing then
        deleted = redis.call("LREM", keyQueueRequeued, 1, messageId)
    elseif messageStatus == EMessagePropertyStatusPending then
        local queueType = redis.call("HGET", keyQueueProperties, EQueuePropertyQueueType)
        if queueType ~= false then
            if queueType == EQueuePropertyQueueTypePriorityQueue then
                deleted = redis.call("ZREM", keyQueuePriorityPending, messageId)
            end
            if queueType == EQueuePropertyQueueTypeFIFOQueue or queueType == EQueuePropertyQueueTypeLIFOQueue then
                deleted = redis.call("LREM", keyQueuePending, 1, messageId)
            end
        end
    end
    if deleted == 1 then
        redis.call("DEL", keyMessage)
        return 'OK'
    end
    return 'MESSAGE_NOT_DELETED'
end

local deleteMessageStatus = 'INVALID_PARAMETERS'

if #ARGV > argvIndexOffset then
    for index in pairs(ARGV) do
        if (index > argvIndexOffset) then
            messageId = ARGV[index]
            keyQueueScheduled = KEYS[keyIndexOffset + 1]
            keyQueueDelayed = KEYS[keyIndexOffset + 2]
            keyQueueRequeued = KEYS[keyIndexOffset + 3]
            keyMessage = KEYS[keyIndexOffset + 4]
            keyQueueProperties = KEYS[keyIndexOffset + 5]
            keyQueuePending = KEYS[keyIndexOffset + 6]
            keyQueueDL = KEYS[keyIndexOffset + 7]
            keyQueueAcknowledged = KEYS[keyIndexOffset + 8]
            keyQueuePriorityPending = KEYS[keyIndexOffset + 9]
            keyIndexOffset = keyIndexOffset + 9
            deleteMessageStatus = deleteMessage()
            if deleteMessageStatus ~= 'OK' then
                break
            end
            updateQueue()
        end
    end
end

return deleteMessageStatus

File: packages/redis-smq/src/lib/message/_/_delete-message.ts
/*
 * Copyright (c)
 * Weyoss <weyoss@protonmail.com>
 * https://github.com/weyoss
 *
 * This source code is licensed under the MIT license found in the LICENSE file
 * in the root directory of this source tree.
 */

import {
  async,
  CallbackEmptyReplyError,
  CallbackInvalidReplyError,
  ICallback,
  IRedisClient,
} from 'redis-smq-common';
import { ELuaScriptName } from '../../../common/redis-client/scripts/scripts.js';
import { redisKeys } from '../../../common/redis-keys/redis-keys.js';
import { EQueueProperty, EQueueType } from '../../queue/index.js';
import {
  MessageInvalidParametersError,
  MessageMessageNotFoundError,
} from '../errors/index.js';
import {
  TMessageDeleteRawResponse,
  IMessageDeleteResponse,
  EMessageProperty,
  EMessagePropertyStatus,
} from '../types/index.js';
import { _getMessage } from './_get-message.js';

function isDeleteMessageResponse(
  reply: unknown,
): reply is TMessageDeleteRawResponse {
  return (
    // type-coverage:ignore-next-line
    typeof reply === 'string' || (Array.isArray(reply) && reply.length === 4)
  );
}

export function _deleteMessage(
  redisClient: IRedisClient,
  messageId: string | string[],
  cb: ICallback<IMessageDeleteResponse>,
): void {
  const keys: string[] = [];
  const argv: (string | number)[] = [];
  const ids = typeof messageId === 'string' ? [messageId] : messageId;
  argv.push(
    EQueueProperty.QUEUE_TYPE,
    EQueueProperty.MESSAGES_COUNT,
    EQueueType.PRIORITY_QUEUE,
    EQueueType.LIFO_QUEUE,
    EQueueType.FIFO_QUEUE,
    EMessageProperty.STATUS,
    EMessagePropertyStatus.PROCESSING,
    EMessagePropertyStatus.ACKNOWLEDGED,
    EMessagePropertyStatus.PENDING,
    EMessagePropertyStatus.SCHEDULED,
    EMessagePropertyStatus.DEAD_LETTERED,
    EMessagePropertyStatus.UNACK_DELAYING,
    EMessagePropertyStatus.UNACK_REQUEUING,
  );

  const deleteResponse: IMessageDeleteResponse = {
    status: 'MESSAGE_NOT_DELETED',
    stats: {
      processed: 0,
      success: 0,
      notFound: 0,
      inProcess: 0,
    },
  };

  const stats = deleteResponse.stats;

  // Skip message retrieval for empty array
  if (ids.length === 0) {
    return cb(null, deleteResponse);
  }

  async.each(
    ids,
    (id, _, done) => {
      _getMessage(redisClient, id, (err, message) => {
        if (err) {
          // If message not found, we'll assume it's already deleted
          if (err instanceof MessageMessageNotFoundError) {
            stats.notFound += 1;
            stats.processed += 1;
            return done();
          }
          return done(err);
        } else if (!message) done(new CallbackEmptyReplyError());
        else {
          const {
            keyQueueScheduled,
            keyQueueDelayed,
            keyQueueRequeued,
            keyQueueProperties,
            keyQueueDL,
            keyQueueAcknowledged,
            keyQueuePriorityPending,
            keyQueuePending,
            keyQueueMessages,
          } = redisKeys.getQueueKeys(
            message.getDestinationQueue(),
            message.getConsumerGroupId(),
          );
          const { keyMessage } = redisKeys.getMessageKeys(id);
          keys.push(
            keyQueueScheduled,
            keyQueueDelayed,
            keyQueueRequeued,
            keyMessage,
            keyQueueProperties,
            keyQueuePending,
            keyQueueDL,
            keyQueueAcknowledged,
            keyQueuePriorityPending,
            keyQueueMessages,
          );
          argv.push(id);
          done();
        }
      });
    },
    (err) => {
      if (err) return cb(err);

      if (keys.length && argv.length) {
        return redisClient.runScript(
          ELuaScriptName.DELETE_MESSAGE,
          keys,
          argv,
          (err, reply: unknown) => {
            if (err) return cb(err);
            if (!isDeleteMessageResponse(reply))
              return cb(new CallbackInvalidReplyError());
            if (typeof reply === 'string') {
              if (reply === 'INVALID_PARAMETERS')
                return cb(new MessageInvalidParametersError());
              return cb(new CallbackInvalidReplyError());
            }
            const [processed, success, notFound, inProcess] = reply;
            stats.processed += processed;
            stats.success += success;
            stats.notFound += notFound;
            stats.inProcess += inProcess;

            deleteResponse.status =
              (stats.processed === stats.success && 'OK') ||
              (stats.success && 'PARTIAL_SUCCESS') ||
              'MESSAGE_NOT_DELETED';

            cb(null, deleteResponse);
          },
        );
      }

      return cb(null, deleteResponse);
    },
  );
}
